src/configuration_server.erl

```diff
@@ -0,0 +1,207 @@
+%%%-------------------------------------------------------------------
+%%% @doc
+%%% Configuration_server finds, reads and stores the configuration for TEP. It presents an interface
+%%% for other processes to access configuration variables.
+%%% @end
+%%%-------------------------------------------------------------------
+-module(configuration_server).
+
+-behaviour(gen_server).
+
+-include("configuration_server.hrl").
+
+%%% Export these internal functions if EUnit is testing this module
+-ifdef(EUNIT).
+
+-export([
+    search_for_config_file_path/2
+]).
+
+-endif.
+
+%% API
+-export([
+    start_link/0,
+    get_default_directories/0,
+    get_default_filenames/0,
+    get_config_file_path/0
+]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).
+
+-define(SERVER, ?MODULE).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+%% @doc Spawns the server and registers the local name (unique)
+-spec start_link() ->
+    {ok, Pid :: pid()} | ignore | {error, Reason :: term()}.
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+%% @doc Returns a list with the default configuration file locations
+-spec get_default_directories() ->
+    {ok, [file:filename_all()]}.
+get_default_directories() ->
+    gen_server:call(configuration_server, get_default_directories).
+
+%% @doc Returns a list with the default names for configuration files
+-spec get_default_filenames() ->
+    {ok, [file:filename_all()]}.
+get_default_filenames() ->
+    gen_server:call(configuration_server, get_default_filenames).
+
+%% @doc Returns the currently used configuration file
+-spec get_config_file_path() ->
+    {ok, file:filename_all()} | {error, Reason :: term()}.
+get_config_file_path() ->
+    gen_server:call(configuration_server, get_config_file_path).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+%% @private
+%% @doc Initializes the server
+-spec init(Args :: term()) ->
+    {ok, State :: #configuration{}} |
+    {ok, State :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term()} | ignore.
+init([]) ->
+    ConfigFilePathSearchResult = search_for_config_file_path(
+        ?TEP_DEFAULT_CONFIG_LOCATIONS,
+        ?TEP_DEFAULT_CONFIG_FILE_NAMES
```

Comment:
I have a problem with the multiple use of these macros. These macros currently expand only into a list, containing a few function calls. This might change. The impact on the execution would be easier to be accounted for, if it is encapsuled in a  dedicated function. This would also reduce the overhead of multiple copies in the source code and therefore also in the compiled modules. This would in my opinion violate the DRY principle.

Reply:
We checked if this has any effect on the compiled code size, but that is negligable. So for now, lets keep using this.